File: pset2/caesar/caesar.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <cs50.h>

char* getPlainText(void);
void encryptMessage(int, char*);
char changeAlphASCII(int, char);

int main(int argc, string argv[])
{
    if (argc != 2) {
        printf("Usage: ./caesar k\n");
        return 1;
    }
    // "No need to check that it’s indeed numeric."
    int key = atoi(argv[1]);
    encryptMessage(key, getPlainText());
    return 0;
}

// Get plaintext from user
char* getPlainText(void)
{
	printf("plaintext: ");
	char* s = get_string();
	
    if (s != NULL) {
        return s;
    }
	else {
	    return "";
	}
}

// encrypting by formula c[i] = (p[i] + k) mod 26
void encryptMessage(int k, char* p)
{
    enum constants { 
        UPP_CASE = 0, LOW_CASE = 1 
    };
    printf("ciphertext: ");
    for (int i = 0, n = strlen(p); i < n; i++) {
        if(isalpha(p[i])) {
            printf("%c", changeAlphASCII(k, p[i]));
        }
        else {
            printf("%c", p[i]);
        }
    }
    printf("\n");
}

// ASCII => alphabet change => ASCII
char changeAlphASCII(int k, char sym)
{
    enum constants {
        ALPHABET = 26, DIFF_CASE = 32, A_ASCII = 65
    };
    int letterСase = 0;
    if(islower(sym)) {
        letterСase = 1;
        // In the case of small letters 65 ('A') and 32 ('A'...'a') are deducted
        sym -= DIFF_CASE;
    }
    // In the case of capital letters 65 ('A') is deducted
    sym = (sym - A_ASCII + k) % ALPHABET;
    sym += A_ASCII;
    // Back to small letters
    if(letterСase) {
        return sym += DIFF_CASE;
    }
    // For capital letters
    return sym;
}

File: pset6/caesar/caesar.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const (
	ALPHABET  = 26
	DIFF_CASE = 32

	A_ASCII = 65
	Z_ASCII = 90

	a_ASCII = 97
	z_ASCII = 122
)

type encryptingString struct {
	// plaintext from user
	plaintext string
	// ciphertext as result
	ciphertext []byte
}

// get plaintext from user
func (es *encryptingString) getPlainText() error {

	fmt.Printf("plaintext: ")

	// constructs a new Reader
	reader := bufio.NewReader(os.Stdin)

	// get entire string of input
	line, err := reader.ReadString('\n')

	if err != nil {
		return err
	}
	// use new string as plaintext
	es.plaintext = line
	return err
}

// encrypting by formula c[i] = (p[i] + k) mod 26
func (es *encryptingString) encryptMessage(k byte) error {

	var err error
	if err = es.getPlainText(); err != nil {
		return err
	}

	fmt.Printf("ciphertext: ")
	// analize each letter
	for _, c := range []byte(es.plaintext) {
		// if is letter
		if c >= A_ASCII && c <= Z_ASCII || c >= a_ASCII && c <= z_ASCII {

			if c, err = changeAlphASCII(k, c); err != nil {
				return err
			}
		}
		// add characters in result
		es.ciphertext = append(es.ciphertext, c)
	}

	return err
}

func main() {

	if len(os.Args) != 2 {

		fmt.Printf("Usage: ./caesar k\n")
		return
	}

	// argument indeed numeric
	key, err := strconv.Atoi(os.Args[1])
	if err != nil {

		fmt.Printf("It’s not indeed numeric.\n")
		return
	}

	es := &encryptingString{}
	if err := es.encryptMessage(byte(key)); err != nil {

		fmt.Println(err)
	} else {
		// result with new line
		fmt.Printf("%s\n", es.ciphertext)
	}
}
